internal/cluster/nodes/tracker_redis.go

```diff
@@ -0,0 +1,202 @@
+	logger logrus.FieldLogger
+
+	picker NodePicker
+
+	client    RedisClient
+	namespace string
+	nodeId    string
+	nodes     map[string]time.Time
+
+	updateInterval time.Duration
+	expiryInterval time.Duration
+}
+
+type RedisClient interface {
+	Subscribe(ctx context.Context, channels ...string) *redis.PubSub
+	Publish(ctx context.Context, channel string, message interface{}) *redis.IntCmd
+}
+
+// NewRedisNodeTracker returns a NodeTracker which tracks nodes in a Redis, and updates the provided NodePicker with
+// lifecycle events.
+//
+// Note that we're not trying to solve the CAP theorem here, if Redis has a bad time, then so do we.
+func NewRedisNodeTracker(
+	logger logrus.FieldLogger,
+	picker NodePicker,
+	redisClient RedisClient,
+	namespace, nodeId string,
+	updateInterval, expiryInterval time.Duration,
+) NodeTracker {
+	return &redisNodeTracker{
+		logger: logger,
+		picker: picker,
+
+		client:    redisClient,
+		namespace: namespace,
+		nodeId:    nodeId,
+		nodes:     make(map[string]time.Time),
+
+		updateInterval: updateInterval,
+		expiryInterval: expiryInterval,
+	}
+}
+
+// Run will track nodes via Redis PubSub until the context is closed.
+func (rnt *redisNodeTracker) Run(ctx context.Context) {
+	clck := clock.FromContext(ctx)
+
+	pubsub := rnt.client.Subscribe(ctx, rnt.namespace)
+	defer pubsub.Close()
+
+	// This forces the SUBSCRIBE to be processed, otherwise we
+	// may send the introduction request before subscribing.
+	if _, err := pubsub.Receive(ctx); err != nil {
+		// If this fails, it will slow down the initial learning of all the hosts,
+		// but it's not critical.  The bigger hazard is that redis itself has failed
+		// and this is just a symptom of a larger problem.
+		rnt.logger.WithError(err).Warning("Subscribe failed")
+	}
+
+	psChan := pubsub.Channel() // Closed when pubsub is Closed
+
+	// Send an immediate heartbeat, this will also solicit other nodes to respond.
+	if err := rnt.sendIntroductionRequest(ctx); err != nil {
+		rnt.logger.WithError(err).Warning("Initial redis check in failed")
+	}
+
+	// Starting the ticker is how we signal to tests that everything is ready to go.
+	ticker := clck.NewTicker(rnt.updateInterval)
+	defer ticker.Stop()
+
+	defer func() {
+		// On shutdown, remove ourselves from other hosts + the NodePicker.
+		ctxExit, cancel := clck.TimeoutContext(context.Background(), 1*time.Second)
+		rnt.sendDrop(ctxExit)
+		cancel()
+		if rnt.nodeId != "" {
+			rnt.picker.Remove(rnt.nodeId)
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			rnt.expireNodes(clck.Now())
+			if err := rnt.sendHeartbeat(ctx); err != nil {
+				rnt.logger.WithError(err).Warning("Failed to check in to redis")
+			}
+		case msg := <-psChan:
+			rnt.handleMessage(ctx, msg.Payload, clck.Now())
+		}
+	}
+}
+
+func (rnt *redisNodeTracker) handleMessage(ctx context.Context, message string, now time.Time) {
+	if message == "" {
+		return
+	}
+	node := message[1:]
```

Comment:
Does this return bytes or runes here?

Reply:
No idea. This isn't a py3 port, we're dealing with hostnames.  Pretty sure they need to be ascii.

Reply:
Spent to much time watching string fuzzing, questioning everything!